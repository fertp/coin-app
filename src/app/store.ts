import type { PreloadedState } from '@reduxjs/toolkit'
import type { ToolkitStore } from '@reduxjs/toolkit/dist/configureStore'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { coincapApi } from '@/services/api'
import assetSlice from '@/modules/guest/slices/assetSlice'
import { favoriteAssetsSlice } from '@/modules/guest/slices/favoriteAssetSlice'

const rootReducer = combineReducers({
  [coincapApi.reducerPath]: coincapApi.reducer,
  asset: assetSlice,
  favoriteAssets: favoriteAssetsSlice.reducer
})

export const setupStore = (preloadedState?: PreloadedState<RootState>): ToolkitStore => {
  return configureStore({
    reducer: {
      [coincapApi.reducerPath]: coincapApi.reducer,
      asset: assetSlice,
      favoriteAssets: favoriteAssetsSlice.reducer
    },
    preloadedState,
    middleware: getDefaultMiddleware => getDefaultMiddleware().concat(coincapApi.middleware)
  })
}

export type RootState = ReturnType<typeof rootReducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']
