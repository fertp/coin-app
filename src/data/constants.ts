export const API_URL = 'https://api.coincap.io/v2'

export const SM_BREAKPOINT = 640
