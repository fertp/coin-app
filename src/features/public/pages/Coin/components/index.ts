export { Title } from "./Title";
export { Stats } from "./Stats"; 
export { Chart } from "./Chart";
export { MarketsTable } from "./MarketsTable";