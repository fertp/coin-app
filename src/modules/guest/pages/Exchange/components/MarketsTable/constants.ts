type ColumnHeaders = Record<string, string>

export const COLUMN_HEADERS: ColumnHeaders = {
  PAIR: 'Pair',
  RATE: 'Rate',
  PRICE: 'Price',
  VOLUME_24HR: 'Volume (24Hr)',
  VOLUME_PERCENT: 'Volume (%)',
  TRADES_24HR: 'Trades (24Hr)',
  STATUS: 'Status'
}
