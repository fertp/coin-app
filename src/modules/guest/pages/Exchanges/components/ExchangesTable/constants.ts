interface ColumnHeaders {
  RANK: 'Rank'
  NAME: 'Name'
  TRADING_PAIRS: 'Trading Pairs'
  TOP_PAIRS: 'Top Pairs'
  VOLUME_24HR: 'Volume (24Hr)'
  TOTAL: 'Total (%)'
  STATUS: 'Status'
  [key: string]: string
}

export const COLUMN_HEADERS: ColumnHeaders = {
  RANK: 'Rank',
  NAME: 'Name',
  TRADING_PAIRS: 'Trading Pairs',
  TOP_PAIRS: 'Top Pairs',
  VOLUME_24HR: 'Volume (24Hr)',
  TOTAL: 'Total (%)',
  STATUS: 'Status'
}
