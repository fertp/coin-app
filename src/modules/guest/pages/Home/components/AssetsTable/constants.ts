export const COLUMN_HEADERS = {
  RANK: 'Rank',
  NAME: 'Name',
  PRICE: 'Price',
  MARKET_CAP: 'Market Cap',
  CHANGE: 'Change'
}
