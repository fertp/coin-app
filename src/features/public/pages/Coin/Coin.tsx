import { FC, useEffect, useState } from "react";
import { Button, Container, Loader } from "../../components";
import { useParams } from "react-router-dom";
import { Asset, AssetHistory, Market } from "@/interfaces/interfaces";
import { getAssetById, getAssetHistory, getMarkets } from "../../services/api";
import { Stats, Title, Chart, MarketsTable } from "./components";

interface AssetData {
  asset: Asset
  history: AssetHistory[]
  markets: Market[]
}

export const Coin:FC = () => {

  const { id } = useParams()
  const [ isLoading, setIsLoading ] = useState<boolean>(true)
  const [ assetData, setAssetData ] = useState<AssetData>()

  useEffect(() => {

    if ( id === undefined ) return

    const abortController = new AbortController()
    const { signal } = abortController

    const fetchAsset = async () => {
      try {
        const [ asset, history, markets ] =  await Promise.all([ 
          getAssetById({ id, signal }),
          getAssetHistory({ id, signal }),
          getMarkets({ id, signal })
        ])
        setAssetData({ asset, history, markets })
        setIsLoading(false)
      }
      catch (error) {
        console.log(error)
      }
    }
    fetchAsset()

    return () => {
      abortController.abort()
    }

  }, [])

  if ( isLoading ) {
    return <Loader color='#ea580c' className='mt-12 mx-auto' />
  }

  return (
    <Container>
      <section className="px-4 sm:px-8 lg:px-0">
        <div className="flex flex-col md:flex-row flex-wrap gap-8 lg:gap-20 md:items-center">
          <Title asset={assetData?.asset} />

          <Stats 
            asset={assetData?.asset} 
            history={assetData?.history}  
          />
        </div>

        <Chart history={assetData?.history} />
      </section>

      <MarketsTable markets={assetData?.markets} />
    </Container>
  )
}