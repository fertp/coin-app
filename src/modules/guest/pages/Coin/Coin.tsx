import { type FC } from 'react'
import { Container, Loader } from '../../components'
import { useParams } from 'react-router-dom'
import { Stats, Title, Chart, MarketsTable, IntervalButtons } from './components'
import { useGetAssetByIdQuery, useGetAssetHistoryQuery } from '@/services/api'
import { useAppSelector } from '@/app/hooks'

export const Coin: FC = () => {
  const { id } = useParams()

  if (id === undefined) {
    return (
      <Loader
        color='#ea580c'
        fullScreen
      />
    )
  }

  const { timeRange } = useAppSelector(state => state.asset)

  // TODO: Optimize this to reduce renders
  const { data: asset, error: assetError, isLoading: isAssetLoading } = useGetAssetByIdQuery(id)
  const { isLoading: isHistoryLoading } = useGetAssetHistoryQuery({
    id,
    timeRange
  })

  if (assetError != null) {
    // TODO: Check status 404
    return (
      <Container>
        <p>Asset not found</p>
      </Container>
    )
  }

  if (isAssetLoading || isHistoryLoading) {
    return (
      <Loader
        color='#ea580c'
        fullScreen
      />
    )
  }

  return (
    <Container>
      <section className='px-4 sm:px-8 lg:px-0'>
        <div className='flex flex-col flex-wrap gap-8 md:flex-row md:items-center lg:gap-20'>
          <Title asset={asset?.data} />

          <Stats id={id} />
        </div>

        <Chart id={id} />

        <IntervalButtons id={id} />
      </section>

      <MarketsTable id={id} />
    </Container>
  )
}
