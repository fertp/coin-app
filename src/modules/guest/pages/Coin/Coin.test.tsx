import type { RenderResult } from '@testing-library/react'
import { vi } from 'vitest'
import { API_URL } from '@/data/constants'
import { server } from '@/test/mock/server'
import {
  act,
  getAllByRole,
  getByRole,
  getByText,
  render,
  screen,
  waitFor,
  waitForElementToBeRemoved
} from '@testing-library/react'
import { rest } from 'msw'
import { Provider } from 'react-redux'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { Coin } from './Coin'
import { store } from '@/app/store'
import fakeData from '@/test/fakeData/fakeAssets.json'
import fakeAssetMarkets from '@/test/fakeData/fakeAssetMarkets.json'
import { Line } from 'react-chartjs-2'
import userEvent from '@testing-library/user-event'

const fakeAsset = fakeData[0]

const Component = (): JSX.Element => {
  const route = `/coins/${fakeAsset?.id}`

  return (
    <Provider store={store}>
      <MemoryRouter initialEntries={[route]}>
        <Routes>
          <Route
            path='/coins/:id'
            element={<Coin />}
          />
        </Routes>
      </MemoryRouter>
    </Provider>
  )
}

const renderByDefault = async (): Promise<RenderResult> => {
  let wrapper
  await act(async () => {
    wrapper = render(<Component />)
  })

  if (wrapper === undefined) throw new Error('Component could not be rendered')

  return wrapper
}

describe('Coin page', () => {
  describe('When the request is failed', () => {
    /**
     * Have no idea what happen with this one.
     * It keeps setted the error handler for other tests.
     * Really weird.
     */
    it('Should display an error message', async () => {
      server.use(
        rest.get(`${API_URL}/assets/:id`, (req, res, ctx) => {
          res.networkError('Failed to connect')
        })
      )

      const errorMessage = /asset not found/i
      await renderByDefault()
      await waitFor(() => {
        expect(screen.getByText(errorMessage)).toBeInTheDocument()
      })
      server.resetHandlers()
    })
  })

  describe('When it mounts by default with a valid id param', () => {
    it('Should render correctly and display a spinner while fetching the data', async () => {
      const { container } = await renderByDefault()

      const spinner = screen.getByRole('status', { name: /loading/i })
      await waitForElementToBeRemoved(spinner)

      expect(container).toMatchSnapshot()
    })

    it('Should render correctly all asset details', async () => {
      vi.mock('react-chartjs-2')
      if ('render' in Line) {
        ;(Line.render as jest.Mock).mockImplementation((props: { 'aria-label': string }) => (
          <img
            src='canvas'
            alt={props['aria-label']}
          />
        ))
      }

      await renderByDefault()

      await waitFor(() => {
        expect(screen.getByAltText(fakeAsset.name)).toBeInTheDocument()
        const heading = screen.getByRole('heading', { level: 1 })
        expect(heading).toHaveTextContent(`${fakeAsset.name} (${fakeAsset.symbol})`)
        expect(screen.getByRole('list', { name: `${fakeAsset.name} stats` })).toBeInTheDocument()
        expect(screen.getByAltText(/last day asset history chart/i)).toBeInTheDocument()
        expect(screen.getByRole('generic', { name: /select an intervaL/i })).toBeInTheDocument()
        expect(screen.getByRole('table', { name: /markets table/i })).toBeInTheDocument()
      })
    })

    it('Should display correctly formatted asset stats', async () => {
      const expectedAssetStats = {
        rank: '101',
        price: '$15.23',
        high: '$15.53',
        low: '$15.11',
        average: '$15.34',
        change: '-0.75%'
      }

      await renderByDefault()

      const stats = screen.getByRole('list', { name: `${fakeAsset.name} stats` })

      Object.keys(expectedAssetStats).forEach(key => {
        expect(getByText(stats, expectedAssetStats[key as keyof typeof expectedAssetStats])).toBeInTheDocument()
      })
    })

    it('Should render correctly a table with all fetched asset markets', async () => {
      await renderByDefault()
      const marketsTable = screen.getByRole('table', { name: /markets table/i })
      const { 1: tableBody } = getAllByRole(marketsTable, 'rowgroup')
      const rows = getAllByRole(tableBody, 'row')
      expect(rows.length).toBe(fakeAssetMarkets.length)
    })

    describe('When user clicks on time range buttons', () => {
      it('Should render an image with corresponding time range history chart', async () => {
        await renderByDefault()
        const buttons = screen.getByRole('generic', { name: /select an interval/i })

        expect(screen.getByRole('img', { name: /last day asset history chart/i })).toBeInTheDocument()

        await userEvent.click(getByRole(buttons, 'button', { name: /last week/i }))
        expect(screen.getByRole('img', { name: /last week asset history chart/i })).toBeInTheDocument()

        await userEvent.click(getByRole(buttons, 'button', { name: /last month/i }))
        expect(screen.getByRole('img', { name: /last month asset history chart/i })).toBeInTheDocument()

        await userEvent.click(getByRole(buttons, 'button', { name: /last six months/i }))
        expect(screen.getByRole('img', { name: /last six months asset history chart/i })).toBeInTheDocument()

        await userEvent.click(getByRole(buttons, 'button', { name: /last year/i }))
        expect(screen.getByRole('img', { name: /last year asset history chart/i })).toBeInTheDocument()
        // screen.debug()
      })
    })
  })
})
