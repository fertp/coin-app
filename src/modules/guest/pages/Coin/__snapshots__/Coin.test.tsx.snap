// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Coin page > When it mounts by default with a valid id param > Should render correctly and display a spinner while fetching the data 1`] = `
<div>
  <div>
    <div
      class="container mx-auto my-8 sm:my-12 md:my-20 xl:max-w-6xl"
    >
      <section
        class="px-4 sm:px-8 lg:px-0"
      >
        <div
          class="flex flex-col flex-wrap gap-8 md:flex-row md:items-center lg:gap-20"
        >
          <div
            class="flex items-center"
          >
            <img
              alt="Decred"
              class="h-20 w-20"
              src="https://static.coincap.io/assets/icons/dcr@2x.png"
            />
            <div
              class="ml-6 sm:ml-8"
            >
              <h1
                class="text-2xl sm:text-3xl"
              >
                Decred
                Â 
                <small
                  class="text-gray-500"
                >
                  (
                  DCR
                  )
                </small>
              </h1>
              <span
                class="mt-2 text-gray-500"
              >
                26 September 2023
              </span>
            </div>
          </div>
          <div>
            <ul
              aria-label="Decred stats"
              class="grid max-w-xs grid-cols-2 gap-x-8 gap-y-4 sm:max-w-sm sm:grid-cols-3 md:max-w-none"
            >
              <li
                class="flex items-center"
                data-name="Rank"
              >
                <span
                  class="mr-4 text-xs uppercase text-gray-600"
                >
                  Rank
                </span>
                <span>
                  101
                </span>
              </li>
              <li
                class="flex items-center"
                data-name="Price"
              >
                <span
                  class="mr-4 text-xs uppercase text-gray-600"
                >
                  Price
                </span>
                <span>
                  $15.23
                </span>
              </li>
              <li
                class="flex items-center"
                data-name="High"
              >
                <span
                  class="mr-4 text-xs uppercase text-gray-600"
                >
                  High
                </span>
                <span>
                  $15.53
                </span>
              </li>
              <li
                class="flex items-center"
                data-name="Low"
              >
                <span
                  class="mr-4 text-xs uppercase text-gray-600"
                >
                  Low
                </span>
                <span>
                  $15.11
                </span>
              </li>
              <li
                class="flex items-center"
                data-name="Average"
              >
                <span
                  class="mr-4 text-xs uppercase text-gray-600"
                >
                  Average
                </span>
                <span>
                  $15.34
                </span>
              </li>
              <li
                class="flex items-center"
                data-name="Change"
              >
                <span
                  class="mr-4 text-xs uppercase text-gray-600"
                >
                  Change
                </span>
                <span>
                  -0.75%
                </span>
              </li>
            </ul>
          </div>
          <div
            class="flex xl:order-last xl:flex-grow xl:justify-end"
          >
            <button
              class="block rounded border border-orange-500 py-1 px-2 font-semibold hover:border-transparent hover:bg-orange-500 hover:text-white  bg-transparent text-orange-700 undefined"
            >
              Add to
               favorites
            </button>
          </div>
        </div>
        <div
          class="mt-8 md:mt-12"
        >
          <img
            alt="last day asset history chart"
            src="canvas"
          />
        </div>
        <div
          aria-label="Select an interval"
          class="mt-4 flex gap-2"
        >
          <button
            aria-label="last day"
            class="block rounded border border-orange-500 py-1 px-2 font-semibold hover:border-transparent hover:bg-orange-500 hover:text-white  bg-orange-500 text-white relative"
          >
            <span
              class="false"
            >
              1d
            </span>
          </button>
          <button
            aria-label="last week"
            class="block rounded border border-orange-500 py-1 px-2 font-semibold hover:border-transparent hover:bg-orange-500 hover:text-white  bg-transparent text-orange-700 relative"
          >
            <span
              class="false"
            >
              1w
            </span>
          </button>
          <button
            aria-label="last month"
            class="block rounded border border-orange-500 py-1 px-2 font-semibold hover:border-transparent hover:bg-orange-500 hover:text-white  bg-transparent text-orange-700 relative"
          >
            <span
              class="false"
            >
              1m
            </span>
          </button>
          <button
            aria-label="last six months"
            class="block rounded border border-orange-500 py-1 px-2 font-semibold hover:border-transparent hover:bg-orange-500 hover:text-white  bg-transparent text-orange-700 relative"
          >
            <span
              class="false"
            >
              6m
            </span>
          </button>
          <button
            aria-label="last year"
            class="block rounded border border-orange-500 py-1 px-2 font-semibold hover:border-transparent hover:bg-orange-500 hover:text-white  bg-transparent text-orange-700 relative"
          >
            <span
              class="false"
            >
              1y
            </span>
          </button>
        </div>
      </section>
      <div
        class="flex justify-center mx-auto mt-24"
      >
        <span
          aria-label="Loading..."
          role="status"
          style="display: inherit; position: relative; width: 60px; height: 60px;"
        >
          <span
            style="position: absolute; height: 60px; width: 60px; background-color: rgb(234, 88, 12); border-radius: 100%; opacity: 0.6; top: 0px; left: 0px; animation-fill-mode: both; animation: react-spinners-BounceLoader-bounce 2.1s 1s infinite ease-in-out;"
          />
          <span
            style="position: absolute; height: 60px; width: 60px; background-color: rgb(234, 88, 12); border-radius: 100%; opacity: 0.6; top: 0px; left: 0px; animation-fill-mode: both; animation: react-spinners-BounceLoader-bounce 2.1s 0s infinite ease-in-out;"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
